### DISCUSSION TITLE
Stator deflections using StructElmer

### DISCUSSION THREAD

**Discussion by vinayPrakash31** [Eomys/pyleecan#551] [Q&A]
Hey there,
I wanted to know if we could already compute the displacement at the stator due to e-NVH using StructElmer method. I have already tried running it at my end but, at the moment, I struggle to see deflections on the stator. Is there any tutorial on the structural coupling that I could refer from or may be some other leads? I believe that different "groups" can not yet be accessed (not yet implemented) in _build_meshsolution.py_. I currently use Pyleecan v1.3.8. Thank you. :)
Regards, Vinay.

**Comment by BonneelP** [comment]
Hello, 

Thank you for your interest in pyleecan ! 

Unfortunately we don't have a tutorial on this model, instead we have a test that runs on my side (https://github.com/Eomys/pyleecan/blob/master/Tests/Simulation/test_StructElmer.py). But there is no plot/check on the results so I don't know how you can interact with the results. I guess that all is in output.struct.meshsolution. You can also have a look at the code of the module here: https://github.com/Eomys/pyleecan/tree/master/pyleecan/Methods/Simulation/StructElmer (run is the main function) and the parameters of the model are detailed here: https://github.com/Eomys/pyleecan/blob/master/pyleecan/Generator/ClassesRef/Simulation/StructElmer.csv

I'm not really familiar with this part of the code, @SebGue do you think you would have some time to write a small tutorial ?

Best regards,
Pierre

**Comment by SebGue** [comment]
Hello,

actually it is only possible to calculate centrifugal force on the rotor. But I guess, since there were a lot of code changes, this feature need some bug fixing as well. If I checked StructElmer, I will write the long overdue tutorial.

Regarding stator displacement, if you are willing to implement such feature, I can give you some hints where to start. I guess a lot of the actual SructElmer code can be reused. One main 'issue' is to figure out how to imply forces in Elmer.

Best regards, Sebastian

**Reply by vinayPrakash31** [reply]
Hello @SebGue,
Has there been any progress on this issue? I would really like to work on the structural part with StructElmer.
Could you please provide some suggestions on where to start from? Thanks.

Best regards,
Vinay

**Reply by SebGue** [reply]
Hello Vinay,

there has been no progress since my last answer :-(
But as it looks now, I have some time to work on this issue the next two weeks.

Are you familiar with Elmer? Maybe you could get to know how to imply forces (and/or calculate eigenmodes) in Elmer case files. 
Meanwhile I will review geometry creation for a proper stator modeling and setup some first test case.

Best regards, Sebastian

**Reply by vinayPrakash31** [reply]
Hey Sebastian,

Thanks for sparing some time on this. I don't know all the intricacies of Elmer, however I'm referring to a well-documented tutorial on induction machines modelling in Elmer [https://www.researchgate.net/publication/317012206_SEMTEC_Report_Elmer_FEM_-_Induction_Machine_Tutorial](url). I will keep you updated with the progress.

Regards, Vinay

**Reply by vinayPrakash31** [reply]
Hey Sebastian,
After reviewing some standard documents/articles on e-machines modelling using Elmer, it seems that the "nodal forces" can be implied on to the stator as a "body force" with appropriate Dirichlet BC using the commands in the SIF file: 
`Body Force 5
	Displacement 1 load dofs = equals "nodal force 1"
	Displacement 2 load dofs = equals "nodal force 2"
End`
and then using their "Linear Elasticity" solver to get the displacement. However, I have often faced singularity issues while solving for displacements from their linear elasticity model.

**Reply by SebGue** [reply]
Hello Vinay,

I saw some SEMTEC publication with the same or similar commands. Now the question is, what will be your input to the structural simluation. I.e. do you plan to do Elmer magnetic simulations to get the nodal forces or is there any other input force you want to use?

On my side, I reviewed and fixed the StructElmer code (on my StructElmer branch). In the tests there are some examples how to run a respective simulation (Tests/Simulation/test_StructElmer.py).

For stator deflection I guess I will create a new class skeleton and add some common code, next.

Do you have a model that you can upload to check the singularity issue?

Best regards, Sebastian

**Reply by vinayPrakash31** [reply]
Hello Sebastian,
I believe I will be using Elmer's nodal input magnetic forces to get stator's displacements. 

Here's the model (Elmer SIF file) that I'm working on: (it's essentially the same model from SEMTECH publication with some changes to take into account the linear elastic solver). In my opinion, the problem is due to the boundary conditions. Could you please have a look at it?
[model.zip](https://github.com/Eomys/pyleecan/files/9617332/model.zip)

Thank you.
BR, Vinay

**Reply by SebGue** [reply]
Hello Vinay,

I had a first look at your files and tried to run 'model.sif' without any resonable result. 

The file structure seems a little chaotic to me. :-) Maybe you can do some cleanup. I guess there are a lot of unused or duplicate files, e.g. in the 'im' folder. So its hard for me to follow.
As an example here is an Elmer case (generated with pyleecan) with a clean structre.  [Elmer_Elasticity.zip](https://github.com/Eomys/pyleecan/files/9633395/Elmer_Elasticity.zip) 

How do you generate the mesh files, i.e. from gmsh file to elmer grid?

Maybe we can reduce the sif to a running minimal electromagnetic case first. Then if we get valid EM results we can add the structural part.

Best regards, Sebastian
